### PR TITLE
[SPARK-47346][PYTHON] Make daemon mode configurable when creating Python planner workers

### PR DESCRIPTION
**Pull request by allisonwang-db**
### What changes were proposed in this pull request?

This PR adds an extra config to env.createPythonWorker to make daemon mode configurable to give more flexibility when creating Python planner workers.

### Why are the changes needed?

To make python workers more flexible.

### Does this PR introduce _any_ user-facing change?

No 

### How was this patch tested?

Existing tests

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by allisonwang-db** [comment]
cc @ueshin @HyukjinKwon

**Comment by HyukjinKwon** [comment]
im fine w/ this change but would defer to @ueshin

File: core/src/main/scala/org/apache/spark/SparkEnv.scala
Comment (allisonwang-db) [reply core/src/main/scala/org/apache/spark/SparkEnv.scala:157]: I removed the default value here and added another method for creating Python workers without this param.

File: core/src/main/scala/org/apache/spark/api/python/StreamingPythonRunner.scala
Comment (allisonwang-db) [inline core/src/main/scala/org/apache/spark/api/python/StreamingPythonRunner.scala:72]: cc @WweiL

Comment (WweiL) [reply core/src/main/scala/org/apache/spark/api/python/StreamingPythonRunner.scala:72]: Thanks for the ping! Let me close this PR then https://github.com/apache/spark/pull/45448

**Review by zhengruifeng — APPROVED** [review]

File: core/src/main/scala/org/apache/spark/SparkEnv.scala
Comment (ueshin) [inline core/src/main/scala/org/apache/spark/SparkEnv.scala:149]: We may want to check the given factory has the same `useDaemonEnabled` value as the argument `useDaemon` to avoid creating an unexpected worker?

File: core/src/main/scala/org/apache/spark/SparkEnv.scala
Comment (ueshin) [inline core/src/main/scala/org/apache/spark/SparkEnv.scala:172]:
```suggestion
    val useDaemon = conf.get(Python.PYTHON_USE_DAEMON)
    createPythonWorker(
      pythonExec, workerModule, daemonModule, envVars, useDaemon)
```

**Review by ueshin — APPROVED** [review]
LGTM, pending tests.

**Comment by ueshin** [comment]
@allisonwang-db Seems like the new test is still failing:

```
[info] - SPARK-47346: cannot create Python worker with different useDaemon flag *** FAILED *** (34 milliseconds)
[info]   org.apache.spark.SparkException: Error from python worker:
[info]   /usr/bin/python3: Error while finding module specification for 'pyspark.daemon' (ModuleNotFoundError: No module named 'pyspark')
[info] PYTHONPATH was:
[info]   /home/runner/work/spark/spark/core/target/scala-2.13/spark-core_2.13-4.0.0-SNAPSHOT.jar
[info] org.apache.spark.SparkException: EOFException occurred while reading the port number from pyspark.daemon's stdout.
[info]   at org.apache.spark.errors.SparkCoreErrors$.eofExceptionWhileReadPortNumberError(SparkCoreErrors.scala:55)
[info]   at org.apache.spark.api.python.PythonWorkerFactory.startDaemon(PythonWorkerFactory.scala:264)
[info]   at org.apache.spark.api.python.PythonWorkerFactory.createThroughDaemon(PythonWorkerFactory.scala:138)
[info]   at org.apache.spark.api.python.PythonWorkerFactory.create(PythonWorkerFactory.scala:104)
[info]   at org.apache.spark.SparkEnv.createPythonWorker(SparkEnv.scala:156)
[info]   at org.apache.spark.SparkEnv.createPythonWorker(SparkEnv.scala:166)
[info]   at org.apache.spark.api.python.PythonWorkerFactorySuite.$anonfun$new$4(PythonWorkerFactorySuite.scala:64)
```

We may need to set up a working Python env for this test somehow?

**Comment by ueshin** [comment]
The failed test seems not related to this PR.

**Comment by ueshin** [comment]
Thanks! merging to master.

**Pull request closed without merging**
